scio-snowflake/src/main/scala/com/spotify/scio/snowflake/SnowflakeIO.scala

```diff
@@ -0,0 +1,134 @@
+      .map(schema => datasourceBeforeSchema.withSchema(schema))
+      .getOrElse(datasourceBeforeSchema)
+  }
+
+  private[snowflake] def buildCsvMapper[T](rowDecoder: RowDecoder[T]): CsvMapper[T] =
+    new CsvMapper[T] {
+      override def mapRow(parts: Array[String]): T = {
+        val unsnowedParts = parts.map {
+          case "\\N" => "" // needs to be mapped to an Option
+          case other => other
+        }.toSeq
+        rowDecoder.unsafeDecode(unsnowedParts)
+      }
+    }
+
+  private[snowflake] def prepareRead[T](
+    snowflakeOptions: SnowflakeOptions,
+    sc: ScioContext
+  )(implicit rowDecoder: RowDecoder[T], coder: Coder[T]): beam.SnowflakeIO.Read[T] =
+    beam.SnowflakeIO
+      .read()
+      .withDataSourceConfiguration(
+        SnowflakeIO.dataSourceConfiguration(snowflakeOptions.connectionOptions)
+      )
+      .withStagingBucketName(snowflakeOptions.stagingBucketName)
+      .withStorageIntegrationName(snowflakeOptions.storageIntegrationName)
+      .withCsvMapper(buildCsvMapper(rowDecoder))
+      .withCoder(CoderMaterializer.beam(sc, coder))
+}
+
+sealed trait SnowflakeIO[T] extends ScioIO[T]
+
+final case class SnowflakeSelect[T](snowflakeOptions: SnowflakeOptions, select: String)(implicit
+  rowDecoder: RowDecoder[T],
+  coder: Coder[T]
+) extends SnowflakeIO[T] {
+
+  override type ReadP = Unit
+  override type WriteP = Unit
+  override val tapT: TapT.Aux[T, Nothing] = EmptyTapOf[T]
+
+  override protected def read(sc: ScioContext, params: ReadP): SCollection[T] =
+    sc.applyTransform(SnowflakeIO.prepareRead(snowflakeOptions, sc).fromQuery(select))
+
+  override protected def write(data: SCollection[T], params: WriteP): Tap[Nothing] =
+    throw new UnsupportedOperationException("SnowflakeSelect is read-only")
+
+  override def tap(params: ReadP): Tap[Nothing] = EmptyTap
+}
+
+final case class SnowflakeTable[T](snowflakeOptions: SnowflakeOptions, table: String)(implicit
+  rowDecoder: RowDecoder[T],
+  rowEncoder: RowEncoder[T],
+  coder: Coder[T]
+) extends SnowflakeIO[T] {
+
+  override type ReadP = Unit
+  override type WriteP = Unit
+  override val tapT: TapT.Aux[T, Nothing] = EmptyTapOf[T]
+
+  override protected def read(sc: ScioContext, params: ReadP): SCollection[T] =
+    sc.applyTransform(SnowflakeIO.prepareRead(snowflakeOptions, sc).fromTable(table))
+
+  override protected def write(data: SCollection[T], params: WriteP): Tap[Nothing] = {
+    data.applyInternal(
+      beam.SnowflakeIO
+        .write[T]()
+        .withDataSourceConfiguration(
+          SnowflakeIO.dataSourceConfiguration(snowflakeOptions.connectionOptions)
+        )
+        .to(table)
+        .withStagingBucketName(snowflakeOptions.stagingBucketName)
+        .withStorageIntegrationName(snowflakeOptions.storageIntegrationName)
+        .withUserDataMapper(new UserDataMapper[T] {
+          override def mapRow(element: T): Array[AnyRef] = rowEncoder.encode(element).toArray
+        })
+    )
+    EmptyTap
```

Comment:
Isn't it possible to create a tap ?

Reply:
Beam `SnowflakeIO` does not give the path of the temp file, and it's just an intermediary file. So my impression is a tap make little sense here (as much as for `JdbcIo` and `Neo4JIO`).